File: solutions/objective-c/strain/1/Strain.m
#import <Foundation/Foundation.h>
#import "Strain.h"
@implementation NSArray (Strain)
- (NSArray *)discard:(BOOL (^)(id  _Nonnull arrayElement))iteratorBlock {
    return [self keep:NO where:iteratorBlock];
}
- (NSArray *)keep:(BOOL)shouldKeep where:(BOOL (^)(id  _Nonnull arrayElement))iteratorBlock {
    NSMutableArray *outputArray = [[NSMutableArray alloc] init];
    for (id element in self) {
        if (iteratorBlock(element) == shouldKeep) {
            [outputArray addObject:element];
        }
    }
    return outputArray;
}
@end

File: solutions/swift/strain/1/Sources/Strain/Strain.swift
extension Array {
    func keep(_ f: (Element) -> Bool) -> [Element] {
        return result(f)
    }

    func discard(_ f: (Element) -> Bool) -> [Element] {
        return result { (Element) -> Bool in !f(Element) }
    }

    private func result(_ f: (Element) -> Bool) -> [Element] {
        var result = [Element]()
        for e in self {
            if f(e) {
                result.append(e)
            }
        }

        return result
    }
}